nearswap/src/twap.rs

```diff
@@ -0,0 +1,553 @@
+
+        if left_index == 0 {
+            let total_observe = u128::try_from(self.observations[last_index].num_of_observations).unwrap();
+            let price1cumu = u128::try_from(self.observations[last_index].price1_cumulative).unwrap();
+            let price2cumu = u128::try_from(self.observations[last_index].price2_cumulative).unwrap();
+            let mean1 = price1cumu / total_observe;
+            let mean2 = price2cumu / total_observe;
+            return (U128(mean1), U128(mean2));
+        } else {
+            let total_observe = u128::try_from(self.observations[last_index].num_of_observations).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].num_of_observations).unwrap();
+            let price1cumu = u128::try_from(self.observations[last_index].price1_cumulative).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].price1_cumulative).unwrap();
+            let price2cumu = u128::try_from(self.observations[last_index].price2_cumulative).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].price2_cumulative).unwrap();
+            let mean1 = price1cumu / total_observe;
+            let mean2 = price2cumu / total_observe;
+            return (U128(mean1), U128(mean2));
+        }
+    }
+
+    pub(crate) fn log_observation(&mut self, pool: PoolInfo) {
+        // price1, price2 calculate
+        let near = u128::try_from(pool.ynear).unwrap();
+        let reserve = u128::try_from(pool.reserve).unwrap();
+        let price1: u128 = near / reserve;
+        let price2: u128 = reserve / near;
+
+        // update current index
+        if self.populated == 0 {
+            self.last_updated_index = self.initialize(env::block_timestamp(), price1, price2);
+            self.populated += 1;
+        } else {
+            self.last_updated_index = self.write(
+                env::block_timestamp(),
+                price1,
+                price2,
+                MAX_LENGTH
+            );
+        }
+
+        // update mean
+        let len: usize;
+        if self.pivoted == true {
+            len = MAX_LENGTH.into();
+        } else {
+            len = self.observations.len();
+        }
+
+        self.mean_1min = self.calculate_mean(Mean::M_1MIN, len);
+
+        self.mean_5min = self.calculate_mean(Mean::M_5MIN, len);
+
+        self.mean_1h = self.calculate_mean(Mean::M_1H, len);
+
+        self.mean_12h = self.calculate_mean(Mean::M_12H, len);
+    }
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct Observation {
+    // timestamp
+    pub block_timestamp: U64,
+    // Number of observations till block_timestamp
+    pub num_of_observations: U128,
+    // cumulative price of token1 till block_timestamp
+    pub price1_cumulative: U128,
+    // cumulative price of token2 till block_timestamp
+    pub price2_cumulative: U128,
```

Comment:
Please use numerical types (`u64, u128`), instead of strings here. This is not efficient: every time you need to convert a string into a number, and then convert back to string when we log a new transaction.
This will eliminate all `u64::try_from` calls.

Reply:
ok